### ISSUE TITLE
Multi thread concurrent bug

### ISSUE THREAD

**Issue by chengnl** [mybatis/memcached-cache#11]
Threads with the first number of cases, the “groupKey” value will be overwritten, leading to key group the loss of information, update again, after the corresponding value of key cannot be updated.

The code in the MemcachedClientWrapper problem:

```
           Set<String> group = getGroup(groupKey);
        group.add(keyString);

        if (LOG.isDebugEnabled()) {
            LOG.debug("Insert/Updating object ("
                    + groupKey
                    + ", "
                    + group
                    + ")");
        }

       storeInMemcached(groupKey, group);
```

"Groups" will be covered in the multi thread concurrent conditions, resulting in the first "keyString" lost：

```
    if (groups == null) {
        if (LOG.isDebugEnabled()) {
            LOG.debug("Group '"
                    + groupKey
                    + "' not previously stored");
        }
        return new HashSet<String>();
    }
```

**Comment by gweisz** [comment]
Potential solution to this: If groups is null use the memcached Add operation.
However there might exist race conditions within the code:

        // add namespace key into memcached
        Set<String> group = getGroup(groupKey);
        if (group == null) {
            group = new HashSet<String>();
        }
        group.add(keyString);

        if (log.isDebugEnabled()) {
            log.debug("Insert/Updating object ("
                    + groupKey
                    + ", "
                    + group
                    + ")");
        }

        storeInMemcached(groupKey, group);

To solve this gets & cas operations could be used.

Makes sense?

**Comment by hazendaz** [comment]
This should be fixed now with #17 and #45
